pkg/redirectpolicy/manager.go

```diff
@@ -0,0 +1,648 @@
+		if err != nil {
+			continue
+		}
+		podInfo := rpm.getPodMetadata(pod, podIPs)
+		if !config.policyConfigSelectsPod(podInfo) {
+			continue
+		}
+		retPods = append(retPods, podInfo)
+	}
+
+	return retPods
+}
+
+// Validates the given policy config for duplicates.
+func (rpm *Manager) isValidConfig(config LRPConfig) error {
+	if config.lrpType == lrpConfigTypeAddr {
+		fe := config.frontendMappingForAddr.feAddr
+		name, ok := rpm.frontendsByHash[fe.Hash()]
+		if ok && config.name != name {
+			return fmt.Errorf("CiliumLocalRedirectPolicy for"+
+				"frontend %v already exists : %v", fe, config.name)
+		}
+	}
+
+	if config.lrpType == lrpConfigTypeSvc {
+		p, ok := rpm.services[*config.serviceID]
+		// Only 1 serviceMatcher policy is allowed for a service name within a namespace.
+		if ok && config.namespace != "" && config.namespace == rpm.policyConfigs[p].namespace {
+			return fmt.Errorf("CiliumLocalRedirectPolicy for"+
+				" service %v already exists in namespace %v", config.serviceID,
+				config.namespace)
+		}
+	}
+
+	return nil
+}
+
+// mapPodBackendToPolicyFrontend generates and populates frontend to the pod
+// backends mapping for the given policy config.
+func (rpm *Manager) mapPodBackendToPolicyFrontend(config *LRPConfig, pod *podMetadata) {
+	if config.lrpType == lrpConfigTypeAddr {
+		rpm.upsertForConfigTypeAddr(config, pod)
+	} else {
+		rpm.upsertForConfigTypeSvc(config, pod)
+	}
+}
+
+// upsertForConfigTypeAddr upserts a service for the policy config frontend.
+func (rpm *Manager) upsertForConfigTypeAddr(config *LRPConfig, pods ...*podMetadata) {
+	var bes4 []backend
+	var bes6 []backend
+
+	// Generate and map pod backends to the policy frontend.  The policy config
+	// is already sanitized with matching backend and frontend port protocols.
+	// We currently don't check which backends are updated before upserting a
+	// a service with the corresponding frontend. This can be optimized when LRPs
+	// are scaled up.
+	for _, pod := range pods {
+		for _, ip := range pod.ips {
+			for _, portProtocol := range config.backendPorts {
+				var portInt uint16
+				if iana.IsSvcName(portProtocol.Port) {
+					if port, ok := pod.namedPorts[portProtocol.Port]; ok {
+						portInt = port.Port
+					}
+				} else {
+					port, _ := strconv.ParseUint(portProtocol.Port, 0, 16)
+					portInt = uint16(port)
+				}
+				beIP := net.ParseIP(ip)
+				if beIP == nil {
+					continue
+				}
+				be := backend{
+					IP: net.ParseIP(ip),
+					L4Addr: lb.L4Addr{
+						Protocol: lb.L4Type(portProtocol.Protocol),
+						Port:     portInt,
+					},
+				}
+				if config.frontendMappingForAddr.feAddr.IP.To4() != nil {
+					if option.Config.EnableIPv4 {
+						bes4 = append(bes4, be)
+					}
+				} else {
+					if option.Config.EnableIPv6 {
+						bes6 = append(bes6, be)
+					}
+				}
+			}
+
+		}
+		if len(bes4) > 0 {
+			rpm.upsertServiceWithBackends(config, config.frontendMappingForAddr, pod.id, bes4)
+		} else if len(bes6) > 0 {
```

Comment:
Are these conditions exclusive, or should the following be applied?
```suggestion
		}
		if len(bes6) > 0 {
```